### DISCUSSION TITLE
numpy.loadtxt converters argument type error

### DISCUSSION THREAD

**Discussion by tincotema** [microsoft/pyright#6729] [Q&A] [answered]
Hi,
when running the minimal example for the numpy.loadtxt function with the converters argument as given in https://numpy.org/doc/1.26/reference/generated/numpy.loadtxt.html,
i get the following error:
`Argument of type "dict[int, (x: Unknown) -> Any]" cannot be assigned to parameter "converters" of type "Mapping[int | str, (str) -> Any] | None" in function "loadtxt"`
i am not sure if this is a bug or intended behaviour.

the code i run:
```
from io import StringIO
import numpy as np
s = StringIO("1.618, 2.296\n3.141, 4.669\n")
conv = {
    0: lambda x: np.floor(float(x)),  # conversion fn for column 0
    1: lambda x: np.ceil(float(x)),  # conversion fn for column 1

}
a =np.loadtxt(s, delimiter=",", converters=conv)
print(a)
```
I am running python 3.11.6
numpy 1.26.2
pyright 1.1.339
on archlilnux

**Comment by erictraut** [comment, marked as answer]
Pyright is working as intended here. As you can see from the error message, the type of `conv` is incompatible with the type of the `converters` parameter. You haven't provided a type declaration for the `conv` variable, so normal type inference rules are used to determine its type based on the value assigned to it. The dictionary expression in your example uses only `int` keys, so its type is inferred as `dict[int, (x: Unknown) -> Any]`. If you want the type to be compatible with the `converters` parameter, you could add a type annotation for the `conv` variable.

```python
conv: dict[int | str, Callable[[str], Any]] = ...
```

This overrides pyright's default type inference logic. Instead, pyright uses [bidirectional type inference](https://microsoft.github.io/pyright/#/type-inference?id=bidirectional-type-inference-expected-types).

**Reply by tincotema** [reply]
Thanks for the quick answer!
